alg/teca_model_segmentation.py

```diff
@@ -69,51 +76,90 @@ def set_pred_name(self, name):
         """
-        if the data need to be transformed in a way then a function
-        could be provided to be applied on the requested data before
-        running it to the model.
+        Set the number of threds per core. Typically 1 gives the best
+        performance for CPU bound applicaitons, however 2 can sometimes
+        be useful. More than 2 will likeley harm performance.
         """
-        if not hasattr(fn, '__call__'):
-            raise TypeError(
-                "ERROR: The provided data transform function"
-                "is not a function"
-                )
-
-        if not args:
-            raise ValueError(
-                "ERROR: The provided data transform function "
-                "must at least have 1 argument -- the data array object to "
-                "apply the transformation on."
-                )
+        if threads_per_core > 4:
+            raise ValueError('Using more than 2 threads per phyiscal core '
+                             'will degrade performance on most architectures.')
+        self.threads_per_core = threads_per_core
 
-        self.transform_fn = fn
-        self.transport_fn_args = args
-
-    def set_num_threads(self, n):
+    def set_thread_pool_size(self, n_requested):
         """
-        torch: Sets the number of threads used for intra-op parallelism on CPU
+        Sets the number of threads used for intra-op parallelism on CPU
         """
-        # n=-1: use default
-        if n != -1:
-            torch.set_num_threads(n)
+        rank = 0
+        n_ranks = 1
+        comm = self.get_communicator()
+        if teca_py.get_teca_has_mpi():
+            rank = comm.Get_rank()
+            n_ranks = comm.Get_size()
 
-    def get_num_threads(self):
+        n_threads = n_requested
+
+        if n_requested > 0:
+            # pass directly to torch
+            torch.set_num_threads(n_requested)
+        else:
+            # detemrmine the number of physical cores are available
+            # on this node, accounting for all MPI ranks scheduled to
+            # run here.
+            try:
+                n_threads, affinity = \
+                     teca_py.thread_util.thread_parameters(comm, -1, 0, 0)
+
+                # make use of hyper-threads
+                n_threads *= self.threads_per_core
+
+                # pass to torch
+                torch.set_num_threads(n_threads)
+
+            except(RuntimeError):
+                # we failed to detect the number of physical cores per MPI rank
+                # if this is an MPI job then fall back to 2 threads per rank and
+                # if not let torch use all (what happens when you do nothing)
+                n_threads = -1
+                comm = self.get_communicator()
+                if teca_py.get_teca_has_mpi() and comm.Get_size() > 1:
+                    torch.set_num_threads(2)
+                    n_threads = 2
+                else:
+                    pass
```

Comment:
Were you intending for the `pass` statement to be a placeholder for other code?  Writing `else: pass` in python is equivalent to not having an else-block at all, which is why I ask.

Reply:
good point, it's not really needed.

Reply:
I've now fixed this and a few other issues flagged by pycodestyle.